pilot/pkg/model/push_status.go

```diff
@@ -196,3 +233,154 @@ func (ps *PushStatus) UpdateMetrics() {
+		for _, d := range rule.Http {
+			for _, m := range d.Match {
+				for i, g := range m.Gateways {
+					if g != IstioMeshGateway {
+						m.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+					}
+				}
+			}
+			for _, w := range d.Route {
+				w.Destination.Host = ResolveShortnameToFQDN(w.Destination.Host, r.ConfigMeta).String()
+			}
+			if d.Mirror != nil {
+				d.Mirror.Host = ResolveShortnameToFQDN(d.Mirror.Host, r.ConfigMeta).String()
+			}
+		}
+		//resolve host in tcp route.destination
+		for _, d := range rule.Tcp {
+			for _, m := range d.Match {
+				for i, g := range m.Gateways {
+					if g != IstioMeshGateway {
+						m.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+					}
+				}
+			}
+			for _, w := range d.Route {
+				w.Destination.Host = ResolveShortnameToFQDN(w.Destination.Host, r.ConfigMeta).String()
+			}
+		}
+		//resolve host in tls route.destination
+		for _, tls := range rule.Tls {
+			for _, m := range tls.Match {
+				for i, g := range m.Gateways {
+					if g != IstioMeshGateway {
+						m.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+					}
+				}
+			}
+			for _, w := range tls.Route {
+				w.Destination.Host = ResolveShortnameToFQDN(w.Destination.Host, r.ConfigMeta).String()
+			}
+		}
+	}
+
+	return out
+}
+
+// GetServiceAttributes retrieves the UID string of a service and parses
+// the hostname into name and namespace. It was likely the most inefficient
+// method in Istio (and probably beyond).
+func (ps *PushStatus) GetServiceAttributes(hostname Hostname) (*ServiceAttributes, error) {
+	// The original code was calling GetService(host) in each registry, and
+	// GetServiceAttributes in the registry having the entry.
+
+	// For consul and most others, it was returning hostname, default, nil
+	// For ServiceEntryStore, returns the same thing - but checks if
+	// service exists, and returns the correct namespace (no UID) - but that's
+	// the namespace where the ServiceEntry was defined, not where
+	// the workload is running
+
+	// For k8s - it just parses the name and returns UID, while checking
+	// for existence
+	name, namespace, err := parseHostname(hostname)
+	if err != nil {
+		return &ServiceAttributes{
+			Name:      hostname.String(),
+			Namespace: "default",
+		}, nil
+	}
+	// No longer making an expensive check to exist - if we're looking it up,
+	// it was returned from a list services.
+	return &ServiceAttributes{
+			Name:      name,
+			Namespace: namespace,
+			UID:       fmt.Sprintf("istio://%s/services/%s", namespace, name),
+		}, nil
+}
+
+// parseHostname extracts service name and namespace from the service hostname
+func parseHostname(hostname Hostname) (name string, namespace string, err error) {
+	parts := strings.Split(hostname.String(), ".")
+	if len(parts) < 2 {
+		err = fmt.Errorf("missing service name and namespace from the service hostname %q", hostname)
+		return
+	}
+	name = parts[0]
+	namespace = parts[1]
+	return
+}
```

Comment:
All existing implementations were setting namespace to 'default'. 
Yes, the simpler/proper way to do this is to have the service attributes in Service, possibly based on the labels ( like we do with service account for consul ). 

My goal in this PR is to revert the scalability breakage introduced by ServiceAttributes - and let people familiar with what it does provide a proper implementation that scales.

Reply:
Yes but this revert will break all other platforms if shipped as it assumes namespace is part of dns. Including service entries. And this will have some strange behavior when we use *.foo in service entries.

Reply:
@costinm Just add those attributes as fields to Service struct. That avoids the method call, and can be safely populated by each service registry independently. There is no need to introduce a regression in the hopes that someone else will fix it later. @rshriram and mixerplugin owners are on-board with that plan.

Reply:
This PR is not attempting to provide service attributes for other platforms - just to preserve roughly the behavior before - all other platforms except k8s returned 'default' as namespace, service as full service and no uid. 

Adding more attributes to Service struct is wrong - it is fine to use labels, we already do that for serviceaccount - but in 1.1 or someone from mixer team. My focus is on fixing the scalability, the ServiceAttributes was badly implemented and can't stand - most of it doesn't make any sense to me, why would it even call ListServices when all it does is parse the name. 

@rshriram  - if you mean on non-k8s to return the 'default' as before, i.e. detect if it is k8s and do the strange parsing only for k8s - I can do that, it makes sense for Service to keep track of the 
config provider that generated it. And I don't mind adding a proper 'istio-namespace' label for 
non-k8s ( or whatever convention we use for service account )

Reply:
Sure that also works.. As kuat says, I was imagining a simple metadata map[string]string in the service struct. Registries can stick whatever they want in that thing. The only reason I resisted adding a "namespace" field in the service was because its very very easy to start using that throughout the code in places where DNS resolution and others are desired (and I end up being the only one to police these things). Sticking it in metadata will hopefully make people think twice before they add these things arbitrary. (wishful thinking :) )